io/io/src/TFileMerger.cxx

```diff
@@ -608,89 +608,28 @@ Bool_t TFileMerger::MergeRecursive(TDirectory *target, TList *sourcelist, Int_t
                      inputs.Delete();
                   }
                }
-            } else if (cl->IsTObject() &&
-                       cl->GetMethodWithPrototype("Merge", "TCollection*,TFileMergeInfo*") ) {
-               // Object implements Merge(TCollection*,TFileMergeInfo*) and has a reflex dictionary ...
+            } else if (cl->IsTObject()
+                  && (cl->GetMethodWithPrototype("Merge", "TCollection*,TFileMergeInfo*")
+                      || cl->GetMethodWithPrototype("Merge", "TCollection*"))) {
 
-               // Check if already treated
                if (alreadyseen) continue;
 
+               // synthesize a method call according to the TObject
                TList listH;
                TString listHargs;
-               listHargs.Form("(TCollection*)0x%lx,(TFileMergeInfo*)0x%lx", (ULong_t)&listH,(ULong_t)&info);
-
-               // Loop over all source files and merge same-name object
-               TFile *nextsource = current_file ? (TFile*)sourcelist->After( current_file ) : (TFile*)sourcelist->First();
-               if (nextsource == 0) {
-                  // There is only one file in the list
-                  Int_t error = 0;
-                  obj->Execute("Merge", listHargs.Data(), &error);
-                  info.fIsFirst = kFALSE;
-                  if (error) {
-                     Error("MergeRecursive", "calling Merge() on '%s' with the corresponding object in '%s'",
-                           obj->GetName(), key->GetName());
-                  }
+               if (cl->GetMethodWithPrototype("Merge", "TCollection*,TFileMergeInfo*")) {
```

Comment:
This is somewhat an expensive operation.  **If** possible, one should avoid calculating twice in a row.

Reply:
Thank you, should be fixed by 7d4ed3c

Reply:
I think the fall through (non-mergeable case) is broken by 7d4ed3c

Reply:
You're 100% right, fixed for real with 4292cd4